haskell/test/SMT.hs

```diff
@@ -0,0 +1,176 @@
+                    { gtAddress = govAddr
+                    }
+                }
+            }
+        }
+
+callEntrypointAndCompare :: TAddress Parameter -> [ContractCall Parameter] -> ContractState -> EmulatedT PureM ()
+callEntrypointAndCompare _ [] _ = pure ()
+callEntrypointAndCompare dao (cc:ccs) cs = do
+  let haskellResult = callBaseDaoHaskell cc cs
+  ligoResult <- callBaseDaoLigo dao cc cs
+
+  -- Note: `fsConfig` can't be compared because using `getStorage` resulted in different
+  -- different lambda.
+  -- Ex: cRejectedProposalSlashValue:
+  -- - from haskell: `Seq (Seq (DROPN (SS (SS SZ))) (PUSH (VNat 0))) Nop`
+  -- - from ligo: ` Seq (AnnPAIR : % %) (Seq (AnnUNPAIR @ @ % %) (Seq (DROPN (SS (SS SZ))) (PUSH (VNat 0))))`
+
+  (   haskellResult & csStorage & fsStorage
+    , haskellResult & csError
+    , cc
+    ) @==
+    ( ligoResult & csStorage & fsStorage
+    -- , ligoResult & csError
+    , []
+    , cc
+    )
+
+  callEntrypointAndCompare dao ccs haskellResult
+
+callBaseDaoLigo
+  :: TAddress Parameter
+  -> ContractCall Parameter
+  -> ContractState
+  -> EmulatedT PureM ContractState
+callBaseDaoLigo dao cc cs = do
+  case (cc & ccAdvanceLevel) of
+    Just lvl -> advanceLevel lvl
+    Nothing -> pure ()
+
+  result <- attempt @NettestFailure $ callEntrypoint cc dao
+
+  let errors :: [(Int, ModelError)] = case result of
+        Right _ -> []
+        Left (NettestFailedWith _ (MM.ExpressionString tval)) ->
+          [(cc & ccIdx, contractErrorToModelError tval)]
+        Left (NettestFailedWith _ (MM.ExpressionPrim (MM.MichelinePrimAp (MM.MichelinePrimitive "Pair") (MM.ExpressionString tval :<| _) _))) ->
+          [(cc & ccIdx, contractErrorToModelError tval)]
+        Left err -> error $ "Unexpected error:" <> show err
+
+
+  newStorage <- getFullStorage (unTAddress dao)
+
+  pure cs { csStorage = newStorage, csError = errors <> (cs & csError)}
+
+callEntrypoint :: MonadNettest caps base m => ContractCall Parameter -> TAddress Parameter -> m ()
+callEntrypoint cc dao = withSender (cc & ccSender) $ case cc & ccParameter of
+  XtzAllowed (Propose p) -> call dao (Call @"Propose") p
+  XtzAllowed (Transfer_contract_tokens p) -> call dao (Call @"Transfer_contract_tokens") p
+  XtzAllowed (Transfer_ownership p) -> call dao (Call @"Transfer_ownership") p
+  XtzAllowed (Accept_ownership p) -> call dao (Call @"Accept_ownership") p
+
+  XtzForbidden (Vote p) -> call dao (Call @"Vote") p
+  XtzForbidden (Flush p) -> call dao (Call @"Flush") p
+  XtzForbidden (Freeze p) -> call dao (Call @"Freeze") p
+  XtzForbidden (Unfreeze p) -> call dao (Call @"Unfreeze") p
+  XtzForbidden (Update_delegate p) -> call dao (Call @"Update_delegate") p
+  XtzForbidden (Drop_proposal p) -> call dao (Call @"Drop_proposal") p
+
+  XtzAllowed (CallCustom _) -> error "Ligo: `CallCustom` is not supported "
+
+-- | Haskell model that mimicks the expected contract behavior
+callBaseDaoHaskell
```

Comment:
Here's what I mean, if we have a state that we check at each step, we can make a simple monad with [transformers](https://hackage.haskell.org/package/transformers):
```haskell
newtype ModelMonad = ExceptT ModelError (State ModelState) ()
```

the "actions" that you'll have to execute from `ContractCall Parameter` can then have the type of `ModelMonad` and this step-by-step function should have the type of:
```haskell
callBaseDaoHaskell :: ContractCall Parameter -> ModelState -> Either ModelError ModelState
```

P.s. I suggest moving this stuff related to the model to a specific `SMT.Model` module

Reply:
Make a lot of sense, I'll make this changes accordingly.